src/audio/dai-zephyr.c

```diff
@@ -0,0 +1,1408 @@
+	if (!group && flags & DAI_CREAT)
+		group = dai_group_alloc();
+
+	if (group) {
+		/* Group might've been previously unused */
+		if (!group->group_id)
+			group->group_id = group_id;
+
+		group->num_dais++;
+	} else {
+		tr_err(&dai_comp_tr, "dai_group_get(): failed to get group_id %u",
+		       group_id);
+	}
+
+	return group;
+}
+
+void dai_group_put(struct dai_group *group)
+{
+        group->num_dais--;
+
+        /* Mark as unused if there are no more DAIs in this group */
+        if (!group->num_dais)
+                group->group_id = 0;
+}
+
+/* Assign DAI to a group */
+int dai_assign_group(struct comp_dev *dev, uint32_t group_id)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+
+	if (dd->group) {
+		if (dd->group->group_id != group_id) {
+			comp_err(dev, "dai_assign_group(), DAI already in group %d, requested %d",
+				 dd->group->group_id, group_id);
+			return -EINVAL;
+		}
+
+		/* No need to re-assign to the same group, do nothing */
+		return 0;
+	}
+
+	dd->group = dai_group_get(group_id, DAI_CREAT);
+	if (!dd->group) {
+		comp_err(dev, "dai_assign_group(), failed to assign group %d",
+			 group_id);
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "dai_assign_group(), group %d num %d",
+		 group_id, dd->group->num_dais);
+
+	/* Register for the atomic trigger event */
+	notifier_register(dev, dd->group, NOTIFIER_ID_DAI_TRIGGER,
+			  dai_atomic_trigger, 0);
+
+	return 0;
+}
+
+struct dai *dai_get(uint32_t type, uint32_t index, uint32_t flags)
+{
+	const struct dai_type_info *dti;
+	struct dai *d;
+
+	tr_info(&dai_comp_tr, "log_ssp_data get_dai_zephyr index: %u", index);
+
+	dti = dai_find_type(type);
+	if (!dti)
+		return NULL; /* type not found */
+
+	for (d = dti->dai_array; d < dti->dai_array + dti->num_dais; d++) {
+		if (d->index == index)
+			return dai_probe(d->drv) ? NULL : d;
+	}
+
+	return NULL;
+}
+
+void dai_put(struct dai *dai)
+{
+	int ret;
+
+	ret = dai_remove(dai->drv);
+	if (ret < 0) {
+		tr_err(&dai_comp_tr, "dai_put_zephyr: index %d failed ret = %d",
+		       dai->index, ret);
+	}
+}
+
+
+int dai_init(struct sof *sof)
+{
+	struct dai *dai;
+	int i;
+
+#if CONFIG_INTEL_SSP
+	const struct dai_type_info *ssp = NULL;
+	char ssp_name[] = "SSP_X";
+#endif
+#if CONFIG_INTEL_ALH
+	const struct dai_type_info *alh = NULL;
+#endif
+
+	for (i = 0; i < ARRAY_SIZE(dti); i++)
+		switch (dti[i].type) {
+#if CONFIG_INTEL_SSP
+		case SOF_DAI_INTEL_SSP:
+			ssp = dti + i;
+			break;
+#endif
+#if CONFIG_INTEL_ALH
+		case SOF_DAI_INTEL_ALH:
+			alh = dti + i;
+			break;
+#endif
+		}
+
+#if CONFIG_INTEL_SSP
+	dai = ssp->dai_array;
+
+	/* init ssp */
+	for (i = 0; i < ssp->num_dais; i++) {
+		/* name varied by zephyr device tree index */
+		dai->index = i;
+		dai->type = SOF_DAI_INTEL_SSP;
+		dai->dma_caps = DMA_CAP_GP_LP | DMA_CAP_GP_HP;
+		dai->dma_dev = DMA_DEV_SSP;
+		ssp_name[4] = (uint8_t)((i % 10UL) + '0');
+		dai->drv = device_get_binding(ssp_name);
+		if (!dai->drv)
+			return -EINVAL;
```

Comment:
you mean I should directly reference the zephyr driver struct? I guess the zephyr way is to call device_get_binding()... the dai struct is not const as I can't put the pointer there with function in struct initialization.

dma stuff is here just for compatibility as they are directly referenced from other parts of sof code. I guess they should in the end come from zpehyr side with dma driver. Or then perhaps dai properties.... not sure

Reply:
No I mean a local data structure that defines all the DAIs and has any local const data we need for them. i.e to translate from topology to DT
```c
static struct dai_desc {
    int topology_type;
    int topology_id;
    const char *name;
   ...
};

static const struct dai_desc dais[] = {
    {TOPOLOGY_DAI_TYPE_SSP, 1, "SSP_1"},
   ...
};
```

Reply:
ok, so if the device_get_binding is ok to call with all methods and not just in the sof boot, this works. I need to ask around...

Reply:
This can still be done at boot, there is just not need to code the data config when we can define it as a table (where the dai_init() really shrinks to loop over table and get the dais).